internal/querynodev2/segments/state/load_state_lock.go

```diff
@@ -0,0 +1,149 @@
+package state
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/cockroachdb/errors"
+)
+
+type loadStateEnum int
+
+// LoadState represent the state transition of segment.
+// LoadStateOnlyMeta: segment is created with meta, but not loaded.
+// LoadStateDataLoading: segment is loading data.
+// LoadStateDataLoaded: segment is full loaded, ready to be searched or queried.
+// LoadStateReleasing: segment is releasing resources.
+// LoadStateReleased: segment is released.
+// LoadStateOnlyMeta -> LoadStateDataLoading -> LoadStateDataLoaded -> LoadStateReleasing -> (LoadStateReleased or LoadStateOnlyMeta)
+const (
+	LoadStateOnlyMeta    loadStateEnum = iota
+	LoadStateDataLoading               // There will be only one goroutine access segment when loading.
+	LoadStateDataLoaded
+	LoadStateReleasing // There will be only one goroutine access segment when releasing.
+	LoadStateReleased
+)
+
+// LoadState is the state of segment loading.
+func (ls loadStateEnum) String() string {
+	switch ls {
+	case LoadStateOnlyMeta:
+		return "meta"
+	case LoadStateDataLoading:
+		return "loading"
+	case LoadStateDataLoaded:
+		return "loaded"
+	case LoadStateReleasing:
+		return "releasing"
+	case LoadStateReleased:
+		return "released"
+	default:
+		return "unknown"
+	}
+}
+
+// NewLoadStateLock creates a LoadState.
+func NewLoadStateLock(state loadStateEnum) *LoadStateLock {
+	if state != LoadStateOnlyMeta && state != LoadStateDataLoaded {
+		panic(fmt.Sprintf("invalid state for construction of LoadStateLock, %s", state.String()))
+	}
+
+	l := &sync.RWMutex{}
+	return &LoadStateLock{
+		RWMutex: l,
+		cv:      sync.Cond{L: l},
+		state:   state,
+	}
+}
+
+// LoadStateLock is the state of segment loading.
+type LoadStateLock struct {
+	*sync.RWMutex
+	cv    sync.Cond
+	state loadStateEnum
+}
+
+// RLockIfNotReleased locks the segment if the state is not released.
+func (ls *LoadStateLock) RLockIfNotReleased() bool {
+	ls.RLock()
+	if ls.state == LoadStateReleased {
+		ls.RUnlock()
+		return false
+	}
+	return true
+}
+
+// StartLoadData starts load segment data
+// Fast fail if segment is not in LoadStateOnlyMeta.
+func (ls *LoadStateLock) StartLoadData() (*LoadStateLockGuard, error) {
+	// only meta can be loaded.
+	ls.cv.L.Lock()
+	defer ls.cv.L.Unlock()
+
+	if ls.state == LoadStateDataLoaded {
+		return nil, nil
+	}
+	if ls.state != LoadStateOnlyMeta {
+		return nil, errors.New("segment is not in LoadStateOnlyMeta, cannot start to loading data")
+	}
+	ls.cv.Broadcast()
+	ls.state = LoadStateDataLoading
```

Comment:
Maybe change state before broad casting? Otherwise some `waitUntilReleasable` may return false positive here

Reply:
state modification can be only seen after Unlock.
So these order of state change and broadcast will be ok.